### ISSUE TITLE
SSL error

### ISSUE THREAD

**Issue by schmurfy** [Yoshiofthewire/CrunchyXBMC#27]
The addon was working fine until yesterday but for some reason today I cannot manage to use it, on screen I get `Session failed: Check Login` As the only item shown and in the logs I get:

```
21:34:04 T:140504511665920  NOTICE: CR: Starting new session
21:34:04 T:140504511665920  NOTICE: CR: makeAPIRequest: get JSON
21:34:04 T:140504511665920  NOTICE: CR: makeAPIRequest: url = https://api.crunchyroll.com/start_session.0.json
21:34:04 T:140504511665920  NOTICE: CR: makeAPIRequest: options = locale=enUS&access_token=XXX&version=2313.8&device_type=com.crunchyroll.iphone&device_id=XXX
21:34:05 T:140504511665920   ERROR: CR: makeAPIRequest: Connection failed: URLError(SSLError(1, u'[SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:590)'),)
21:34:05 T:140504511665920   ERROR: CR: Error starting new session. Error message: Connection failed: <class 'urllib2.URLError'>, URLError(SSLError(1, u'[SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:590)'),)
```

I looked into this more and this specific error seems linked to the lack of SNI support in http libraries but I have no clue why it worked before unless crunchyroll changed something.

Does anyone else got this ?

I am running kodi 15.0 (it worked one day with this release before I got this)

**Comment by michaeltandy** [comment]
I've encountered the same problem with Kodi 
```
21:47:48 T:139747300767744  NOTICE: Starting Kodi (15.0 Git:2f34a0c). Platform: Linux x86 64-bit
21:47:48 T:139747300767744  NOTICE: Using Release Kodi x64 build
21:47:48 T:139747300767744  NOTICE: Kodi compiled Jul 22 2015 by GCC 4.9.2 for Linux x86 64-bit version 3.19.8 (201480)
21:47:48 T:139747300767744  NOTICE: Running on Ubuntu 15.04, kernel: Linux x86 64-bit version 3.19.0-23-generic
21:47:48 T:139747300767744  NOTICE: FFmpeg statically linked, version: 2.6.3-kodi-2.6.3-Isengard
[...]
21:48:10 T:139745275406080  NOTICE: [PLUGIN] 'Crunchyroll: version 2.1.2' initialized!
21:48:10 T:139745275406080  NOTICE: CR: Starting new session
21:48:10 T:139745275406080  NOTICE: CR: makeAPIRequest: get JSON
21:48:10 T:139745275406080  NOTICE: CR: makeAPIRequest: url = https://api.crunchyroll.com/start_session.0.json
21:48:10 T:139745275406080  NOTICE: CR: makeAPIRequest: options = locale=enUS&access_token=<snip>&version=2313.8&device_type=com.crunchyroll.iphone&device_id=<snip>
21:48:10 T:139745275406080   ERROR: CR: makeAPIRequest: Connection failed: URLError(SSLError(1, u'[SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:590)'),)
21:48:10 T:139745275406080   ERROR: CR: Error starting new session. Error message: Connection failed: <class 'urllib2.URLError'>, URLError(SSLError(1, u'[SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:590)'),)
21:48:10 T:139745753560832  NOTICE: Thread BackgroundLoader start, auto delete: false
21:48:12 T:139745719990016   ERROR: Unable to find plugin
21:48:12 T:139747300767744   ERROR: GetDirectory - Error getting plugin://
21:48:12 T:139747300767744   ERROR: CGUIMediaWindow::GetDirectory(plugin://) failed
```

Looking in Wireshark, seems like the response to the client hello is SSL alert message 0x02 (fatal) 0x80 (internal_error) - but I can connect to api.crunchyroll.com OK using my web browser. Kodi and Firefox alike are willing to negotiate TLSv1.2 and cipher TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 (0xc02b)

Among other differences, firefox, sends the Server Name Indication (SNI) TLS extension while Kodi doesn't. Indeed, if I direct Firefox to https://104.27.149.74/ - the IP I resolved api.crunchyroll.com to - I get a 'ssl_error_internal_error_alert' error

I think we need to convince the plugin to send the SNI TLS extension.

**Comment by michaeltandy** [comment]
Inspired by https://bugs.launchpad.net/ubuntu/+source/python-requests/+bug/1299075 and http://stackoverflow.com/a/26392858/1367431 I ran 

sudo apt-get install python-pip python-openssl python-pyasn1 libffi-dev python-dev
pip install ndg-httpsclient
pip install pyopenssl
pip install ndg-httpsclient

At this point the SNI header still isn't being sent, but Kodi seems to have started working again despite its absence. Maybe a config change on the other end? Or perhaps those commands changed some other TLS config and made things work. Difficult to say, now.

In summary, I didn't knowingly fix anything and it's kinda bullshit that in this day and age python doesn't come with SNI as standard.

**Comment by schmurfy** [comment]
At least I know I am not the only one ^^
I found an issue open on urllib2 (shown in both our logs) about the lack of SNI support, on the so thread the answer you linked mention installing urllib3, maybe that's what happened for you (it may have been pulled as a dependency).

I will give it a try this evening.

**Comment by Yoshiofthewire** [comment]
I was able to replicate this on my RPi2 last night.
Then the problem went away,
CR changed something recently that broke the code, so I hacked around the issue in 2.1.2.
This looked to be the same sort of issue.

**Comment by schmurfy** [comment]
I forgot to report here but for some reasons it worked fine without me doing anything when I tried and has worked fine since then Oo

**Comment by Yoshiofthewire** [comment]
Site Issue, closing

**Comment by carnivash** [comment]
Seem the problem came back, it's been days that i'm getting the error, tried every solution i could find but no luck.

**Comment by Neogrey** [comment]
Are your sure this is the same issue ? . Crunchyroll had some problems with ddos attacks in the last few days, maybe this coursed your issue. Currently works fine for me.

**Comment by carnivash** [comment]
I think so, i tried everything, delete user data, reinstall plugin, reinstall kodi (deleting user data), i even tried to install kodi + crunchyroll on a newly formatted windows machine.

**Comment by Yoshiofthewire** [comment]
I know the github tree is burned.  There was a patch that did work that I need to revert ... while keeping the additional language files.  But I have trouble testing, since I stopped subscribing.

**Comment by carnivash** [comment]
How can i help? i have no problem in sharing my personal Crunchyroll credentials if that's what you need to fix.

**Comment by Neogrey** [comment]
@Yoshiofthewire or, I can send you an "Premium+ Guest Pass - 48 Hours"

And maybe you can take a look at this: 
https://github.com/Yoshiofthewire/CrunchyXBMC/issues/43 https://github.com/Yoshiofthewire/CrunchyXBMC/issues/46

**Comment by carnivash** [comment]
I found these documents but no luck either, premium guest pass here it is: http://www.crunchyroll.com/guest_pass?code=WPKAVEPPQ3F

**Comment by carnivash** [comment]
Still getting the same error, no progress so far?

**Comment by lrenaud** [comment]
@Yoshiofthewire I'd be interested in assisting further if you can provide any insight into how you went about doing your development for the plugin. I attempted to setup a development environment that was a little more interactive to work on the bug I filed, but the Kodi wiki was sending me to out of date web sites with dead links so I lost motivation.

If you're unable to do active work to test things, and can provide some basics as to your debugging/development flow 'd be willing to take another stab at my bug. I got somewhat tired with hacking by trial and error.

**Comment by Yoshiofthewire** [comment]
Well it started as picking up someone else's abandoned plugin.
Then when that plugin became unworkable, I ported over a Plex plugin, which went from pulling the video off the web to acting like a PS3.  A Premium membership was required with this change, and the dropping of the older Vizio TV app.
